File: src/mupstudio/schema/selection.py
"""Which cells something applies to.

Every part of a model that is not uniform has to say where it applies: a well
field, a fixed-head edge, a conductivity zone, an initial water composition.
Those are the same question asked four times, so they get one answer here and
each of them holds a `CellSelection`.

There are three honest ways to point at cells, and a modeller uses all three on
the same model:

  - by index, when the grid is small enough to count and the cells form a block
  - by an explicit list, when they were clicked in the viewport
  - by a shape, when the geometry came from GIS and the cells follow from it

The third is the one that survives a grid change. A river drawn as a line stays
the river after the grid is refined; a list of cell indices taken from that line
becomes wrong the moment a cell size changes, silently. So a shape selection
stores the shape and resolves to cells at compile time, never the other way
round.
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, Field, field_validator

from mupstudio.schema.common import Id


class CellRange(BaseModel):
    """Cells named by index, as the outer product of layers, rows and columns.

    Indices are 1-based to match how MODFLOW input reads, since anyone checking
    this against a listing file is counting from one.
    """

    kind: Literal["cells"] = "cells"
    layers: list[int] = Field(min_length=1)
    rows: list[int] = Field(min_length=1)
    columns: list[int] = Field(min_length=1)

    @field_validator("layers", "rows", "columns")
    @classmethod
    def _one_based(cls, value: list[int]) -> list[int]:
        if any(index < 1 for index in value):
            raise ValueError("that is not a cell; indices start at 1, as MODFLOW input does")
        return value


class CellList(BaseModel):
    """Cells named one at a time.

    What clicking in the viewport produces. A block would be a `CellRange`;
    this is for the cases that are not a block — a diagonal fault trace, a
    handful of wells, the cells left over after a boundary was trimmed.
    """

    kind: Literal["list"] = "list"
    indices: list[tuple[int, int, int]] = Field(
        default_factory=list,
        description="(layer, row, column) triples, 1-based",
    )

    @field_validator("indices")
    @classmethod
    def _one_based(cls, value: list[tuple[int, int, int]]) -> list[tuple[int, int, int]]:
        if any(index < 1 for cell in value for index in cell):
            raise ValueError("that is not a cell; indices start at 1, as MODFLOW input does")
        # Duplicates would write the same MODFLOW record twice, which some
        # packages sum and others reject. Removing them here means neither
        # happens by accident, and clicking a cell twice is not a data error.
        return list(dict.fromkeys(value))


class ShapeSelection(BaseModel):
    """Cells picked out by an imported shape.

    The shape is the truth and the cells are derived, so refining the grid
    re-derives them instead of leaving a stale list behind. This is the only
    selection that stays correct across a regrid.
    """

    kind: Literal["shape"] = "shape"
    source: Id = Field(description="Id of the imported data source to intersect with")
    layers: list[int] = Field(default_factory=lambda: [1], min_length=1)
    rule: Literal["intersects", "centroid"] = Field(
        default="intersects",
        description=(
            "intersects takes every cell the shape touches, centroid only cells whose "
            "centre it covers. Centroid suits an area, intersects suits a line."
        ),
    )
    buffer: float = Field(
        default=0.0,
        ge=0.0,
        description=(
            "Widen the shape by this distance first, in model length units. A river "
            "drawn as a line with a 50 m buffer catches its floodplain too."
        ),
    )

    @field_validator("layers")
    @classmethod
    def _one_based(cls, value: list[int]) -> list[int]:
        if any(index < 1 for index in value):
            raise ValueError("that is not a layer; indices start at 1, as MODFLOW input does")
        return value


CellSelection = Annotated[
    CellRange | CellList | ShapeSelection,
    Field(discriminator="kind"),
]


def cells(layers: list[int], rows: list[int], columns: list[int]) -> CellRange:
    """Shorthand, since a block of cells is the common case in a small model."""
    return CellRange(layers=layers, rows=rows, columns=columns)


def out_of_range(selection: CellSelection | None, *, nlay: int, nrow: int, ncol: int) -> str | None:
    """The first index outside the grid, described, or None if all fit.

    An index typo is worth catching here rather than in a MODFLOW listing file,
    where it appears as a cell number with no hint of which package put it
    there. A shape selection can only be wrong about layers: which rows and
    columns it covers is the grid's answer, not the user's.
    """
    limits = (("layer", nlay), ("row", nrow), ("column", ncol))

    if selection is None:
        return None

    if selection.kind == "shape":
        return _first_bad(selection.layers, "layer", nlay)

    if selection.kind == "list":
        for cell in selection.indices:
            for value, (axis, limit) in zip(cell, limits, strict=True):
                if not 1 <= value <= limit:
                    return f"{axis} {value}, but the grid has {limit} (indices start at 1)"
        return None

    for axis, limit in limits:
        found = _first_bad(getattr(selection, f"{axis}s"), axis, limit)
        if found:
            return found
    return None


def _first_bad(indices: list[int], axis: str, limit: int) -> str | None:
    for index in indices:
        if not 1 <= index <= limit:
            return f"{axis} {index}, but the grid has {limit} (indices start at 1)"
    return None


def describe(selection: CellSelection | None) -> str:
    """A short phrase for a selection, for a list row or a validation message."""
    if selection is None:
        return "everywhere it applies"
    if selection.kind == "cells":
        count = len(selection.layers) * len(selection.rows) * len(selection.columns)
        return f"{count} cell{'s' if count != 1 else ''} by index"
    if selection.kind == "list":
        count = len(selection.indices)
        return f"{count} cell{'s' if count != 1 else ''} picked"
    layers = ", ".join(str(index) for index in selection.layers)
    return f"from {selection.source} in layer {layers}"

File: frontend/src/pages/editor/selection.ts
/**
 * Which cells something applies to, in the browser.
 *
 * Mirrors the schema's three ways of naming cells. The distinction that matters
 * for the screen is that the first two are lists of numbers the user owns, and
 * the third is a question the server answers — so a shape selection cannot be
 * counted or drawn without asking, and the other two can.
 */

export type CellTriple = [number, number, number];

export interface CellRange {
  kind: "cells";
  layers: number[];
  rows: number[];
  columns: number[];
}

export interface CellList {
  kind: "list";
  indices: CellTriple[];
}

export interface ShapeSelection {
  kind: "shape";
  source: string;
  layers: number[];
  rule: "intersects" | "centroid";
  buffer: number;
}

export type CellSelection = CellRange | CellList | ShapeSelection;

export const SELECTION_MODES = [
  {
    kind: "cells" as const,
    label: "Range",
    hint: "Layers, rows and columns by number. Every combination of them is taken, so 2 layers and 3 rows is 6 cells.",
  },
  {
    kind: "list" as const,
    label: "Picked",
    hint: "Cells clicked in the viewport. Use this for anything that is not a block: a diagonal, a scatter of wells, an edge with a notch in it.",
  },
  {
    kind: "shape" as const,
    label: "From data",
    hint: "Cells under an imported shape. This is the only one that stays right after the grid is rebuilt, because the shape is kept and the cells are worked out again.",
  },
];

/** A one-based cell, from what the viewport's picker returns. */
export function cellFromPick(picked: { layer: number; cell: number }, columns: number): CellTriple {
  return [picked.layer + 1, Math.floor(picked.cell / columns) + 1, (picked.cell % columns) + 1];
}

export function sameCell(a: CellTriple, b: CellTriple): boolean {
  return a[0] === b[0] && a[1] === b[1] && a[2] === b[2];
}

/**
 * Add a cell, or remove it if it is already there.
 *
 * Clicking is how a selection is both built and corrected, and a click that
 * only ever adds makes the mistake unfixable without leaving the viewport.
 */
export function toggleCell(indices: CellTriple[], cell: CellTriple): CellTriple[] {
  const without = indices.filter((item) => !sameCell(item, cell));
  return without.length === indices.length ? [...indices, cell] : without;
}

/** The cells a selection names, where that can be known without the server. */
export function localCells(selection: CellSelection | null): CellTriple[] | null {
  if (!selection) return null;
  if (selection.kind === "list") return selection.indices;
  if (selection.kind === "cells") {
    const out: CellTriple[] = [];
    for (const layer of selection.layers) {
      for (const row of selection.rows) {
        for (const column of selection.columns) out.push([layer, row, column]);
      }
    }
    return out;
  }
  // A shape has to be resolved against the grid, which only the server can do.
  return null;
}

export function describeSelection(selection: CellSelection | null): string {
  const cells = localCells(selection);
  if (cells) return `${cells.length} cell${cells.length === 1 ? "" : "s"}`;
  if (selection?.kind === "shape") return `from ${selection.source}`;
  return "everywhere";
}

/** An empty selection of a given kind, keeping what carries across. */
export function emptySelection(
  kind: CellSelection["kind"],
  previous: CellSelection | null,
  firstSource: string,
): CellSelection {
  // Layers survive a mode change because "which layers is this in" is a
  // decision independent of how the footprint was chosen, and retyping it
  // every time the mode is switched is pure friction.
  const layers = previous
    ? previous.kind === "list"
      ? [...new Set(previous.indices.map((cell) => cell[0]))].sort((a, b) => a - b)
      : previous.layers
    : [1];

  if (kind === "cells") return { kind: "cells", layers, rows: [1], columns: [1] };
  if (kind === "list") return { kind: "list", indices: [] };
  return { kind: "shape", source: firstSource, layers, rule: "intersects", buffer: 0 };
}

/** Parse "1, 3, 5-8" into [1, 3, 5, 6, 7, 8]. */
export function parseIndexList(text: string, limit: number): number[] {
  const found = new Set<number>();

  for (const part of text.split(",")) {
    const trimmed = part.trim();
    if (!trimmed) continue;

    // A range is how anyone writes "the whole west edge" without typing
    // ninety numbers, and it is what every other modelling tool accepts.
    const range = /^(\d+)\s*-\s*(\d+)$/.exec(trimmed);
    if (range) {
      const [from, to] = [Number(range[1]), Number(range[2])];
      for (let index = Math.min(from, to); index <= Math.max(from, to); index++) {
        if (index >= 1 && index <= limit) found.add(index);
      }
      continue;
    }

    const value = Math.round(Number(trimmed));
    if (Number.isFinite(value) && value >= 1 && value <= limit) found.add(value);
  }

  return [...found].sort((a, b) => a - b);
}

/** The inverse, collapsing runs back into ranges so the field stays readable. */
export function formatIndexList(indices: number[]): string {
  if (indices.length === 0) return "";
  const sorted = [...indices].sort((a, b) => a - b);
  const parts: string[] = [];

  let start = sorted[0];
  let previous = sorted[0];

  const flush = () => {
    if (start === previous) parts.push(String(start));
    else if (previous === start + 1) parts.push(`${start}, ${previous}`);
    else parts.push(`${start}-${previous}`);
  };

  for (const value of sorted.slice(1)) {
    if (value === previous + 1) {
      previous = value;
      continue;
    }
    flush();
    start = previous = value;
  }
  flush();

  return parts.join(", ");
}
